numpy/core/src/umath/logical_gufuncs.c.src

```diff
@@ -0,0 +1,607 @@
+        }                                       \
+    }
+
+/* -------------------------------------------------------------------------- */
+/* create the family of functions using a template  */
+
+/**begin repeat
+* #TYPE = byte,ubyte,short,ushort,int,uint,long,ulong,longlong,ulonglong,float,
+*         double,longdouble,cfloat,cdouble,clongdouble#
+* #OP_PREFIX = R,R,R,R,R,R,R,R,R,R,R,R,R,PyArray_C,PyArray_C,PyArray_C#
+*/
+
+/**begin repeat1
+* #OP = EQ,NE,LT,LE,GT,GE#
+* #OPNAME = equal,not_equal,less,less_equal,greater,greater_equal#
+*/
+
+/**begin repeat2
+* #ALL_OR_ANY = all,any#
+*/
+
+static void base_@TYPE@_@ALL_OR_ANY@_@OPNAME@
+        (char **args, npy_intp *dimensions,
+         npy_intp* steps, void* data)
+{
+    npy_intp i, n;
+    npy_intp N = dimensions[0], I = dimensions[1];
+    char *a_n = args[0], *b_n = args[1], *c_n = args[2];
+    npy_intp a_N = steps[0], b_N = steps[1], c_N = steps[2];
+    npy_intp a_I = steps[3], b_I = steps[4];
+
+    char *a_i, *b_i;
+    npy_@TYPE@ a, b;
+
+    for (n = 0; n < N; n++) {
+        a_i = a_n;
+        b_i = b_n;
+
+        @ALL_OR_ANY@_INNER_LOOP(@OP_PREFIX@@OP@, npy_@TYPE@)
+
+        a_n += a_N;
+        b_n += b_N;
+        c_n += c_N;
+    }
+}
+
+/**end repeat2**/
+/**end repeat1**/
+/**end repeat**/
+
+
+/* -------------------------------------------------------------------------- */
+/* vectorized versions*/
+
+#define BLOCK_SIZE 1024
+#define ALIGNMENT_SIZE 32
+#define ALIGNED(p) ((unsigned long)p & ALIGNMENT_SIZE)
+
+/*
+*  The core array must be contiguous, ie the core dimension steps must
+*  be equal to the sizeof the type
+*
+*  TODO: all contiguous signed and unsigned integer types could be handled by
+*  a single vectorized all_equal function.  Besides there should be at least 
+*  one simple C function without src templating.
+*/
+/**begin repeat
+* #TYPE = float,double,int#
+**/
+void simd_core_contiguous_@TYPE@_all_equal(
+    char **args, npy_intp *dimensions, npy_intp* steps, void* data)
+{
+    npy_intp i, n, j;
+    npy_intp N = dimensions[0], I = dimensions[1];
+    char *a_n = args[0], *b_n = args[1], *c_n = args[2];
+    npy_intp a_N = steps[0], b_N = steps[1], c_N = steps[2];
+
+    unsigned int true_count;
+    @TYPE@ *x, *y;
+
+    for (n = 0; n < N; n++) {
+        *((npy_bool *)c_n) = NPY_TRUE;
+        x = (@TYPE@ *)a_n;
+        y = (@TYPE@ *)b_n;
+
+        // get x aligned, helps compiler vectorize.  y can use unaligned
+        // loads. do while so at least the first element is checked
+        // before checking an entire chunk
+        i = 0;
+        do {
+            if ((*x) == (*y)) {
+                x++;
+                y++;
+                i++;
+            } else {
+                *((npy_bool *)c_n) = NPY_FALSE;
+                i = I;
+            }
+        } while (!ALIGNED(x) & (i<I));
+
+        // main loop in chunks with auto vectorize simd instructions
+        for ( ; i<(I-BLOCK_SIZE) ; i+=BLOCK_SIZE) {
```

Comment:
I agree it isn't required for a hand written loop with intrinsics or asm.  However for what I tried it did seem required to get the compiler to do a decent job at autovectorization.  Probably super dependent on compiler.  I was mostly using gcc 4.9 which is a bit old.